freqtrade/exchange/exchange.py

```diff
@@ -815,6 +821,20 @@ def _get_params(self, ordertype: str, leverage: float, time_in_force: str = 'gtc
             params.update({param: time_in_force})
         return params
 
+    def _amount_to_contract_size(self, pair: str, amount: float):
+
+        if ('contractSize' in self.markets[pair]):
+            return amount / self.markets[pair]['contractSize']
+        else:
```

Comment:
This is wrong, and will break spot functionality on binance (and probably most other exchanges, even though i didn't bother to verify) with "unsupported operands - division by float and NoneType", as binance markets do contain ` 'contractSize': None`.

That's not wrong per se (from an information perspective) (and having the key there but empty is better than having a missing key in my opinion) - and we'll always have to have proper handling for this anyway.

It points to a 2nd problem though - which is that this is not tested.

The 3rd problem is described more in detail in the other comment - However - due to the way python handles floats, you can end up with `3.3000000000000003` - which is WAY over precision the exchange supports.

Reply:
> assuming i wanna order a 100 contracts of XRP (ignore the price for now) - and it's contractSize is 15 .. so i'd be ordering `100/15=6.666` contracts - wouldn't that mean i'll not be getting 100 contracts, but 6.666 instead?

So you wouldn't want to order 100 contracts, you would want to order 0.0005 BTC or something, but you can't place an order in btc amount, you have to place it in contract size amount, so we have to find out how many contracts would be worth 0.0005 BTC. If the contract size is 0.0001, then you would place an order for 5 contracts which is amount(0.0005)/contract_size(0.0001)

> The binance case makes this worse - assuming i order 1 BTC - it'd have to be `1 / 0.0001` - which is 10.000 contracts ... ??

Binance is the normal case, where you place an order for 0.0001 if you want 0.0001 BTC

Reply:
yeah, binance was a typo - i was actually using the numbers in your original post for this :laughing:

Reply:
Do you still think is wrong?

Reply:
The code didn't change - so neither did my opinion on this.

It MIGHT be correct for futures  (it's not and will fail in case of None) - but as you're calling this without conditions, it'll simply break freqtrade for spot markets depending on what ccxt returns.

Reply:
the correct way to test for this is in `_get_contract_size` - which is however not used here - which makes this method wrong (and the overall picture a little confusing)

Reply:
I think I fixed the problem